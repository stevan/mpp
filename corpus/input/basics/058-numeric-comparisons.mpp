# Numeric comparison operators
my $eq = $a == $b;
my $ne = $x != $y;
my $lt = $count < 10;
my $gt = $size > 100;
my $le = $index <= $max;
my $ge = $value >= $min;

# Spaceship operator (numeric comparison)
my $spaceship = $num1 <=> $num2;
my $numeric_sort = $a <=> $b;

# Chained comparisons
my $in_range = $min <= $value && $value <= $max;
my $valid = $age >= 18 && $age < 65;

# Complex numeric comparisons
if ($score > 90 && $attendance >= 95) { }
if ($balance < 0 || $overdraft > $limit) { }

# Comparisons in ternary
my $grade = $score >= 90 ? "A" : $score >= 80 ? "B" : "C";

# Comparisons with expressions
my $check = ($x + $y) > ($z * 2);
my $test = ($a - $b) <= 0;