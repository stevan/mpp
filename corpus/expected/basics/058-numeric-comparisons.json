[
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$eq"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "==",
      "left": {
        "type": "Variable",
        "name": "$a"
      },
      "right": {
        "type": "Variable",
        "name": "$b"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$ne"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "!=",
      "left": {
        "type": "Variable",
        "name": "$x"
      },
      "right": {
        "type": "Variable",
        "name": "$y"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$lt"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "<",
      "left": {
        "type": "Variable",
        "name": "$count"
      },
      "right": {
        "type": "Number",
        "value": "10"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$gt"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ">",
      "left": {
        "type": "Variable",
        "name": "$size"
      },
      "right": {
        "type": "Number",
        "value": "100"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$le"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "<=",
      "left": {
        "type": "Variable",
        "name": "$index"
      },
      "right": {
        "type": "Variable",
        "name": "$max"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$ge"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ">=",
      "left": {
        "type": "Variable",
        "name": "$value"
      },
      "right": {
        "type": "Variable",
        "name": "$min"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$spaceship"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "<=>",
      "left": {
        "type": "Variable",
        "name": "$num1"
      },
      "right": {
        "type": "Variable",
        "name": "$num2"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$numeric_sort"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "<=>",
      "left": {
        "type": "Variable",
        "name": "$a"
      },
      "right": {
        "type": "Variable",
        "name": "$b"
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$in_range"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "&&",
      "left": {
        "type": "BinaryOp",
        "operator": "<=",
        "left": {
          "type": "Variable",
          "name": "$min"
        },
        "right": {
          "type": "Variable",
          "name": "$value"
        }
      },
      "right": {
        "type": "BinaryOp",
        "operator": "<=",
        "left": {
          "type": "Variable",
          "name": "$value"
        },
        "right": {
          "type": "Variable",
          "name": "$max"
        }
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$valid"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "&&",
      "left": {
        "type": "BinaryOp",
        "operator": ">=",
        "left": {
          "type": "Variable",
          "name": "$age"
        },
        "right": {
          "type": "Number",
          "value": "18"
        }
      },
      "right": {
        "type": "BinaryOp",
        "operator": "<",
        "left": {
          "type": "Variable",
          "name": "$age"
        },
        "right": {
          "type": "Number",
          "value": "65"
        }
      }
    }
  },
  {
    "type": "If",
    "condition": {
      "type": "BinaryOp",
      "operator": "&&",
      "left": {
        "type": "BinaryOp",
        "operator": ">",
        "left": {
          "type": "Variable",
          "name": "$score"
        },
        "right": {
          "type": "Number",
          "value": "90"
        }
      },
      "right": {
        "type": "BinaryOp",
        "operator": ">=",
        "left": {
          "type": "Variable",
          "name": "$attendance"
        },
        "right": {
          "type": "Number",
          "value": "95"
        }
      }
    },
    "thenBlock": [],
    "elseIfClauses": []
  },
  {
    "type": "If",
    "condition": {
      "type": "BinaryOp",
      "operator": "||",
      "left": {
        "type": "BinaryOp",
        "operator": "<",
        "left": {
          "type": "Variable",
          "name": "$balance"
        },
        "right": {
          "type": "Number",
          "value": "0"
        }
      },
      "right": {
        "type": "BinaryOp",
        "operator": ">",
        "left": {
          "type": "Variable",
          "name": "$overdraft"
        },
        "right": {
          "type": "Variable",
          "name": "$limit"
        }
      }
    },
    "thenBlock": [],
    "elseIfClauses": []
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$grade"
    },
    "initializer": {
      "type": "Ternary",
      "condition": {
        "type": "BinaryOp",
        "operator": ">=",
        "left": {
          "type": "Variable",
          "name": "$score"
        },
        "right": {
          "type": "Number",
          "value": "90"
        }
      },
      "trueExpr": {
        "type": "String",
        "value": "\"A\""
      },
      "falseExpr": {
        "type": "Ternary",
        "condition": {
          "type": "BinaryOp",
          "operator": ">=",
          "left": {
            "type": "Variable",
            "name": "$score"
          },
          "right": {
            "type": "Number",
            "value": "80"
          }
        },
        "trueExpr": {
          "type": "String",
          "value": "\"B\""
        },
        "falseExpr": {
          "type": "String",
          "value": "\"C\""
        }
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$check"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": ">",
      "left": {
        "type": "BinaryOp",
        "operator": "+",
        "left": {
          "type": "Variable",
          "name": "$x"
        },
        "right": {
          "type": "Variable",
          "name": "$y"
        }
      },
      "right": {
        "type": "BinaryOp",
        "operator": "*",
        "left": {
          "type": "Variable",
          "name": "$z"
        },
        "right": {
          "type": "Number",
          "value": "2"
        }
      }
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Variable",
      "name": "$test"
    },
    "initializer": {
      "type": "BinaryOp",
      "operator": "<=",
      "left": {
        "type": "BinaryOp",
        "operator": "-",
        "left": {
          "type": "Variable",
          "name": "$a"
        },
        "right": {
          "type": "Variable",
          "name": "$b"
        }
      },
      "right": {
        "type": "Number",
        "value": "0"
      }
    }
  }
]